core/watcher/eventqueue/key.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package eventqueue
+
+import (
+	"github.com/juju/errors"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/core/changestream"
+	"github.com/juju/juju/core/watcher"
+)
+
+// KeyWatcher watches for changes to single database table row.
+// Any time the identified row changes, a notification is emitted.
+type KeyWatcher struct {
+	*BaseWatcher
+
+	out       chan struct{}
+	tableName string
+	keyValue  string
+}
+
+// NewKeyWatcher returns a new watcher that receives changes from the input
+// base watcher's db/queue when a specific database table row changes.
+func NewKeyWatcher(base *BaseWatcher, tableName string, keyValue string) *KeyWatcher {
+	w := &KeyWatcher{
+		BaseWatcher: base,
+		out:         make(chan struct{}),
+		tableName:   tableName,
+		keyValue:    keyValue,
+	}
+
+	w.tomb.Go(w.loop)
+	return w
+}
+
+// Changes returns the channel on which notifications
+// are sent when the watched database row changes.
+func (w *KeyWatcher) Changes() watcher.NotifyChannel {
+	return w.out
+}
+
+func (w *KeyWatcher) loop() error {
+	opt := changestream.FilteredNamespace(w.tableName, changestream.All, func(e changestream.ChangeEvent) bool {
```

Comment:
maybe it's a naive question, but wouldn't it be cheaper to filter events by querying the change stream with proper sql instead of retrieving everything and filtering with go code?

Reply:
100000 workers individually polling and running queries to notice any changes will not scale well. Instead, if we have one worker reading from the changelog for changes allows us to scale much more efficiently.

Reply:
ohhh yes of course, the worker funnels every access to db. Thanks!